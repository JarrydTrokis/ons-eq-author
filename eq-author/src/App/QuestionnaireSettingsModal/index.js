import React from "react";
import PropTypes from "prop-types";
import styled from "styled-components";
import gql from "graphql-tag";
import { noop } from "lodash";

import { useQuestionnaire } from "components/QuestionnaireContext";

import DialogHeader from "components/Dialog/DialogHeader";
import { Message } from "components/Dialog/DialogMessage";
import Modal from "components/modals/Modal";
import QuestionnaireMeta from "./QuestionnaireMeta";

const defaultQuestionnaire = {
  title: "",
  description: "",
  surveyId: "",
  theme: "default",
  navigation: false,
};

const Title = styled.h2`
  margin-left: 0em 0em 0em 0.1em;
  padding-top: 0em;
  padding-bottom: 1em;
  top: 0;
`;

const StyledModal = styled(Modal)`
  .Modal {
    width: 40em;
    padding: 24px;
    padding-top: 0;
  }
`;

const QuestionnaireSettingsModal = ({
  isOpen,
  onClose,
  onSubmit,
  confirmText,
  canEditType,
}) => {
  const { questionnaire } = useQuestionnaire();
  return (
    <StyledModal
      isOpen={isOpen}
      onClose={onClose}
      data={{ test: "questionnaire-settings-modal" }}
    >
      <DialogHeader>
        <Message>
          <Title>New questionnaire</Title>
        </Message>
      </DialogHeader>
      <QuestionnaireMeta
        questionnaire={questionnaire || defaultQuestionnaire}
        onCancel={onClose}
        onSubmit={onSubmit}
        onUpdate={noop}
        confirmText={confirmText}
        canEditType={canEditType}
      />
    </StyledModal>
  );
};

QuestionnaireSettingsModal.propTypes = {
  onSubmit: PropTypes.func,
  isOpen: PropTypes.bool,
  onClose: PropTypes.func,
  confirmText: PropTypes.string.isRequired,
  canEditType: PropTypes.bool,
};

QuestionnaireSettingsModal.defaultProps = {
  questionnaire: defaultQuestionnaire,
  canEditType: true,
};

QuestionnaireSettingsModal.fragments = {
  QuestionnaireSettingsModal: gql`
    fragment QuestionnaireSettingsModal on Questionnaire {
      ...Questionnaire
    }
    ${QuestionnaireMeta.fragments.Questionnaire}
  `,
};

export default QuestionnaireSettingsModal;
