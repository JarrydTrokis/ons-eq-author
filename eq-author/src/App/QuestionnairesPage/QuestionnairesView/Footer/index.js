import React from "react";
import styled from "styled-components";
import PropTypes from "prop-types";

import { colors } from "constants/theme";

import PaginationNav from "./PaginationNav";

const TableFooter = styled.div`
  display: flex;
  padding-top: 1em;
  align-items: center;
`;

const Results = styled.div`
  color: ${colors.textLight};
  font-size: 0.9em;
  position: absolute;
`;

const Footer = ({
  countOnPage,
  totalCount,
  pageCount,
  currentPageIndex,
  onPageChange,
}) => (
  <TableFooter>
    <Results>
      Showing {countOnPage} of {totalCount}
    </Results>
    <PaginationNav
      currentPageIndex={currentPageIndex}
      pageCount={pageCount}
      onPageChange={onPageChange}
    />
  </TableFooter>
);

Footer.propTypes = {
  countOnPage: PropTypes.number.isRequired,
  totalCount: PropTypes.number.isRequired,
  pageCount: PropTypes.number.isRequired,
  currentPageIndex: PropTypes.number.isRequired,
  onPageChange: PropTypes.func.isRequired,
};

export default Footer;
