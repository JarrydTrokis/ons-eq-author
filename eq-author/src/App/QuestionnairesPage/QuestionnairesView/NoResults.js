import React from "react";
import styled from "styled-components";
import PropTypes from "prop-types";

import Button from "components/buttons/Button";
import { colors } from "constants/theme";

import flag from "./flag.svg";

const Wrapper = styled.div`
  padding: 6em 0 10em;
  text-align: center;
`;

const Title = styled.h2`
  font-size: 1.3em;
  font-weight: 500;
  margin: 0 0 0.5em;

  &::before {
    content: url(${flag});
    display: block;
  }
`;

const Text = styled.p`
  margin: 0 0 1em;
  color: ${colors.textLight};
`;

const NoResults = ({ onCreateQuestionnaire }) => (
  <Wrapper>
    <Title>No questionnaires found</Title>
    <Text>{"You don't have access to any questionnaires yet."}</Text>
    <Button
      onClick={onCreateQuestionnaire}
      variant="primary"
      data-test="create-questionnaire"
    >
      Create a questionnaire
    </Button>
  </Wrapper>
);

NoResults.propTypes = {
  onCreateQuestionnaire: PropTypes.func.isRequired,
};

export default NoResults;
