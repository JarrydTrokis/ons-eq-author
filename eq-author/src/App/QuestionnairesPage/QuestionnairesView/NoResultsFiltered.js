import React from "react";
import styled from "styled-components";
import PropTypes from "prop-types";

import { colors } from "constants/theme";

import iconAlert from "./icon-alert.svg";

const Alert = styled.div`
  background: white;
  border-radius: 4px;
  border: 1px solid #ccc;
  padding: 5em;
  text-align: center;
  color: #666;
  display: flex;
  flex-direction: column;
  justify-content: center;
  align-items: center;

  &::before {
    display: inline-block;
    content: url(${iconAlert});
    width: 2em;
    height: 2em;
    margin-bottom: 0.5em;
  }
`;

const AlertTitle = styled.h1`
  font-size: 1.1em;
  margin: 0 0 0.5em;
  font-weight: bold;
  color: ${colors.text};
`;

const AlertText = styled.p`
  font-size: 1em;
  margin: 0;
  font-weight: normal;
`;

const NoResultsFiltered = ({ searchTerm }) => (
  <Alert>
    <AlertTitle>{`No results found for '${searchTerm}'`}</AlertTitle>
    <AlertText>Please check the questionnaire exists</AlertText>
  </Alert>
);

NoResultsFiltered.propTypes = {
  searchTerm: PropTypes.string.isRequired,
};

export default NoResultsFiltered;
