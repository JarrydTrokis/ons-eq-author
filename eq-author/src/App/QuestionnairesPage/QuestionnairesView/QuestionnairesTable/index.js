import React, { useState } from "react";
import PropTypes from "prop-types";
import { propType } from "graphql-anywhere";
import gql from "graphql-tag";

import { SORT_ORDER } from "constants/sort-order.js";
import * as Headings from "constants/table-headings";

import QuestionnaireTable from "components/QuestionnaireTable";
import Panel from "components/Panel";

import tableHeadings from "components/QuestionnaireTable/TableHeadings";

import { useQuestionnaireLockingModal } from "components/modals/QuestionnaireLockingModal";

const enabledRows = [
  Headings.TITLE,
  Headings.OWNER,
  Headings.CREATED,
  Headings.MODIFIED,
  Headings.PERMISSIONS,
  Headings.LOCKED,
  Headings.STARRED,
  Headings.ACTIONS,
];

const QuestionnairesTable = ({
  questionnaires,
  onDeleteQuestionnaire,
  onDuplicateQuestionnaire,
  onSortQuestionnaires,
  onReverseSort,
  autoFocusId,
  sortColumn,
  sortOrder,
}) => {
  const [targetQuestionnaire, setTargetQuestionnaire] = useState({});
  const {
    trigger: triggerLockModal,
    component: LockModal,
  } = useQuestionnaireLockingModal(targetQuestionnaire);

  const handleLock = (questionnaire) => {
    setTargetQuestionnaire(questionnaire);
    triggerLockModal();
  };

  return (
    <Panel>
      <QuestionnaireTable
        onSortClick={onSortQuestionnaires}
        onReverseClick={onReverseSort}
        sortOrder={sortOrder}
        currentSortColumn={sortColumn}
        tableHeadings={tableHeadings}
        questionnaires={questionnaires}
        autoFocusId={autoFocusId}
        onDeleteQuestionnaire={onDeleteQuestionnaire}
        onDuplicateQuestionnaire={onDuplicateQuestionnaire}
        handleLock={handleLock}
        enabledHeadings={enabledRows}
        clickable={false}
      />
      <LockModal />
    </Panel>
  );
};

QuestionnairesTable.fragments = {
  QuestionnaireDetails: gql`
    fragment QuestionnaireDetails on Questionnaire {
      id
      title
      shortTitle
      displayName
      starred
      locked
      createdAt
      updatedAt
      createdBy {
        id
        name
        email
        displayName
      }
      permission
      publishStatus
    }
  `,
};

QuestionnairesTable.propTypes = {
  questionnaires: PropTypes.arrayOf(
    propType(QuestionnairesTable.fragments.QuestionnaireDetails)
  ),
  onDeleteQuestionnaire: PropTypes.func.isRequired,
  onDuplicateQuestionnaire: PropTypes.func.isRequired,
  onSortQuestionnaires: PropTypes.func.isRequired,
  onReverseSort: PropTypes.func.isRequired,
  autoFocusId: PropTypes.string,
  sortColumn: PropTypes.string.isRequired,
  sortOrder: PropTypes.oneOf([SORT_ORDER.ASCENDING, SORT_ORDER.DESCENDING]),
};

export default QuestionnairesTable;
