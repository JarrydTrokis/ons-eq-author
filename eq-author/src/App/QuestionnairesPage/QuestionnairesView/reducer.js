import { chunk, clamp } from "lodash";

const PAGE_SIZE = 16;

const calculateAutoFocusId = (questionnaires, deletedQuestionnaire) => {
  const possibleNextIndex =
    questionnaires.indexOf(
      questionnaires.find(q => q.id === deletedQuestionnaire.id)
    ) + 1;

  // If the last one is being removed then focus the one before that
  const nextIndex =
    possibleNextIndex > questionnaires.length - 1
      ? questionnaires.length - 2
      : possibleNextIndex;

  // We have to set focusedId to undefined when there are no
  // questionnaires left
  return (questionnaires[nextIndex] || {}).id;
};

export const buildInitialState = questionnaires => state => {
  const pages = chunk(questionnaires, PAGE_SIZE);
  const currentPageIndex = clamp(state.currentPageIndex, 0, pages.length - 1);
  return {
    ...state,
    questionnaires,
    pages,
    currentPageIndex,
    currentPage: pages[currentPageIndex],
  };
};

export const ACTIONS = {
  CHANGE_PAGE: "CHANGE_PAGE",
  SET_QUESTIONNAIRES: "SET_QUESTIONNAIRES",
  DELETE_QUESTIONNAIRE: "DELETE_QUESTIONNAIRE",
};

const reducer = (state, action) => {
  switch (action.type) {
    case ACTIONS.CHANGE_PAGE: {
      const currentPageIndex = clamp(action.payload, 0, state.pages.length - 1);
      return {
        ...state,
        currentPageIndex,
        currentPage: state.pages[currentPageIndex],
        autoFocusId: null,
      };
    }
    case ACTIONS.SET_QUESTIONNAIRES:
      return buildInitialState(action.payload)(state);
    case ACTIONS.DELETE_QUESTIONNAIRE:
      return {
        ...state,
        autoFocusId: calculateAutoFocusId(state.questionnaires, action.payload),
      };
    default:
      return state;
  }
};

export default reducer;
