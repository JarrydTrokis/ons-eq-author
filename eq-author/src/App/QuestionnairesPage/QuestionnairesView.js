import React, { PureComponent } from "react";
import PropTypes from "prop-types";
import { propType } from "graphql-anywhere";
import styled from "styled-components";

import { isEmpty } from "lodash";

import QuestionnairesTable from "./QuestionnairesTable";

import { colors } from "constants/theme";

import Button from "components/buttons/Button";
import QuestionnaireSettingsModal from "App/QuestionnaireSettingsModal";
import flag from "./flag.svg";

const Header = styled.div`
  margin: 1em 0;
  display: flex;
  z-index: 1;
  align-items: center;
  justify-content: flex-end;
`;

const NoResults = styled.div`
  padding: 6em 0 10em;
  text-align: center;
`;

const NoResultsTitle = styled.h2`
  font-size: 1.3em;
  font-weight: 500;
  margin: 0 0 0.5em;

  &::before {
    content: url(${flag});
    display: block;
  }
`;

const NoResultsText = styled.p`
  margin: 0 0 1em;
  color: ${colors.textLight};
`;

class QuestionnairesView extends PureComponent {
  state = {
    isModalOpen: false,
  };

  handleModalOpen = () => this.setState({ isModalOpen: true });
  handleModalClose = () => this.setState({ isModalOpen: false });

  render() {
    const {
      questionnaires,
      onDeleteQuestionnaire,
      onDuplicateQuestionnaire,
      onCreateQuestionnaire,
    } = this.props;

    return (
      <>
        {isEmpty(questionnaires) ? (
          <NoResults>
            <NoResultsTitle>No questionnaires found</NoResultsTitle>
            <NoResultsText>
              {"You don't have access to any questionnaires yet."}
            </NoResultsText>
            <Button
              onClick={this.handleModalOpen}
              variant="primary"
              data-test="create-questionnaire"
            >
              Create a questionnaire
            </Button>
          </NoResults>
        ) : (
          <>
            <Header>
              <Button
                onClick={this.handleModalOpen}
                primary
                data-test="create-questionnaire"
              >
                Create questionnaire
              </Button>
            </Header>
            <QuestionnairesTable
              questionnaires={questionnaires}
              onDeleteQuestionnaire={onDeleteQuestionnaire}
              onDuplicateQuestionnaire={onDuplicateQuestionnaire}
            />
          </>
        )}
        <QuestionnaireSettingsModal
          isOpen={this.state.isModalOpen}
          onClose={this.handleModalClose}
          onSubmit={questionnaire => onCreateQuestionnaire(questionnaire)}
          confirmText="Create"
        />
      </>
    );
  }
}

QuestionnairesView.propTypes = {
  questionnaires: PropTypes.arrayOf(
    propType(QuestionnairesTable.fragments.QuestionnaireDetails)
  ),
  onCreateQuestionnaire: PropTypes.func.isRequired,
  onDeleteQuestionnaire: PropTypes.func.isRequired,
  onDuplicateQuestionnaire: PropTypes.func.isRequired,
};

export default QuestionnairesView;
