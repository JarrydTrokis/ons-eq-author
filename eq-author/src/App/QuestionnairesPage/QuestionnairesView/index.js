import React, { useState, useRef } from "react";
import PropTypes from "prop-types";
import { propType } from "graphql-anywhere";
import styled from "styled-components";
import { isEmpty } from "lodash";

import QuestionnaireSettingsModal from "App/QuestionnaireSettingsModal";
import Button from "components/buttons/Button";

import NoResults from "./NoResults";
import QuestionnairesTable from "./QuestionnairesTable";
import Footer from "./Footer";

import reducer, { buildInitialState, ACTIONS } from "./reducer";
import usePersistedReducer from "./usePersistedReducer";

export const STORAGE_KEY = "questionnaire-list-settings";
const STORED_KEYS = ["currentPageIndex"];

const Header = styled.div`
  margin: 1em 0;
  display: flex;
  z-index: 1;
  align-items: center;
  justify-content: flex-end;
`;

const QuestionnairesView = ({
  questionnaires,
  onDeleteQuestionnaire,
  onDuplicateQuestionnaire,
  onCreateQuestionnaire,
}) => {
  const [isModalOpen, setIsModalOpen] = useState(false);

  const handleModalOpen = () => setIsModalOpen(true);
  const handleModalClose = () => setIsModalOpen(false);

  const questionnairesRef = useRef(questionnaires);

  const [state, dispatch] = usePersistedReducer(
    STORAGE_KEY,
    STORED_KEYS,
    reducer,
    { currentPageIndex: 0 },
    buildInitialState(questionnairesRef.current)
  );

  if (questionnaires !== questionnairesRef.current) {
    dispatch({
      type: ACTIONS.SET_QUESTIONNAIRES,
      payload: questionnaires,
    });
    questionnairesRef.current = questionnaires;
  }

  const handleDeleteQuestionnaire = questionnaire => {
    dispatch({
      type: ACTIONS.DELETE_QUESTIONNAIRE,
      payload: questionnaire,
    });
    onDeleteQuestionnaire(questionnaire.id);
  };

  if (isEmpty(state.questionnaires)) {
    return (
      <>
        <NoResults onCreateQuestionnaire={handleModalOpen} />
        <QuestionnaireSettingsModal
          isOpen={isModalOpen}
          onClose={handleModalClose}
          onSubmit={onCreateQuestionnaire}
          confirmText="Create"
        />
      </>
    );
  }

  return (
    <>
      <Header>
        <Button
          onClick={handleModalOpen}
          primary
          data-test="create-questionnaire"
        >
          Create questionnaire
        </Button>
      </Header>
      <QuestionnairesTable
        questionnaires={state.currentPage}
        onDeleteQuestionnaire={handleDeleteQuestionnaire}
        onDuplicateQuestionnaire={onDuplicateQuestionnaire}
        autoFocusId={state.autoFocusId}
      />
      <Footer
        countOnPage={state.currentPage.length}
        totalCount={state.questionnaires.length}
        pageCount={state.pages.length}
        currentPageIndex={state.currentPageIndex}
        onPageChange={newPage =>
          dispatch({ type: ACTIONS.CHANGE_PAGE, payload: newPage })
        }
      />
      <QuestionnaireSettingsModal
        isOpen={isModalOpen}
        onClose={handleModalClose}
        onSubmit={onCreateQuestionnaire}
        confirmText="Create"
      />
    </>
  );
};

QuestionnairesView.propTypes = {
  questionnaires: PropTypes.arrayOf(
    propType(QuestionnairesTable.fragments.QuestionnaireDetails)
  ),
  onCreateQuestionnaire: PropTypes.func.isRequired,
  onDeleteQuestionnaire: PropTypes.func.isRequired,
  onDuplicateQuestionnaire: PropTypes.func.isRequired,
};

export default QuestionnairesView;
