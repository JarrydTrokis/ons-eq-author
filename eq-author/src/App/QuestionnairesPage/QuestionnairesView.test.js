import React from "react";
import { shallow } from "enzyme";
import QuestionnaireSettingsModal from "App/QuestionnaireSettingsModal";
import QuestionnairesView from "App/QuestionnairesPage/QuestionnairesView";

describe("QuestionnairesPage/QuestionnairesView", () => {
  it("allows Modals to be open and closed", () => {
    const wrapper = shallow(
      <QuestionnairesView
        onDeleteQuestionnaire={jest.fn()}
        onCreateQuestionnaire={jest.fn()}
        onDuplicateQuestionnaire={jest.fn()}
      />
    );

    wrapper.find("[data-test='create-questionnaire']").simulate("click");
    expect(wrapper.find(QuestionnaireSettingsModal).prop("isOpen")).toBe(true);

    wrapper.find(QuestionnaireSettingsModal).simulate("close");
    expect(wrapper.find(QuestionnaireSettingsModal).prop("isOpen")).toBe(false);
  });

  it("creates questionnaire after submission", () => {
    const onCreateQuestionnaire = jest.fn();
    const wrapper = shallow(
      <QuestionnairesView
        onDeleteQuestionnaire={jest.fn()}
        onCreateQuestionnaire={onCreateQuestionnaire}
        onDuplicateQuestionnaire={jest.fn()}
      />
    );

    wrapper.find(QuestionnaireSettingsModal).simulate("submit");
    expect(onCreateQuestionnaire).toHaveBeenCalled();
  });

  it("should render questionnaires table when given questionnaires", () => {
    const wrapper = shallow(
      <QuestionnairesView
        onDeleteQuestionnaire={jest.fn()}
        onCreateQuestionnaire={jest.fn()}
        onDuplicateQuestionnaire={jest.fn()}
        questionnaires={[
          {
            id: "1",
            title: "Foo",
            displayName: "Foo",
            shortTitle: "Test title",
            createdAt: "2017/01/02",
            createdBy: {
              id: "2",
              name: "Alan",
            },
          },
        ]}
      />
    );
    expect(wrapper).toMatchSnapshot();
  });
});
