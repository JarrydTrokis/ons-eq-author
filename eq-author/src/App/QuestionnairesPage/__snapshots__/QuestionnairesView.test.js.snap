// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`QuestionnairesPage/QuestionnairesView should render questionnaires table when given questionnaires 1`] = `
<Fragment>
  <QuestionnairesView__Header>
    <Button
      data-test="create-questionnaire"
      onClick={[Function]}
      primary={true}
      type="button"
      variant="primary"
    >
      Create questionnaire
    </Button>
  </QuestionnairesView__Header>
  <UnconnectedQuestionnairesTable
    onDeleteQuestionnaire={[MockFunction]}
    onDuplicateQuestionnaire={[MockFunction]}
    questionnaires={
      Array [
        Object {
          "createdAt": "2017/01/02",
          "createdBy": Object {
            "id": "2",
            "name": "Alan",
          },
          "displayName": "Foo",
          "id": "1",
          "shortTitle": "Test title",
          "title": "Foo",
        },
      ]
    }
  />
  <QuestionnaireSettingsModal
    canEditType={true}
    confirmText="Create"
    isOpen={false}
    onClose={[Function]}
    onSubmit={[Function]}
    questionnaire={
      Object {
        "description": "",
        "navigation": false,
        "surveyId": "",
        "theme": "default",
        "title": "",
      }
    }
  />
</Fragment>
`;
